compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -649,29 +649,334 @@ abstract class GenJSCode extends plugins.PluginComponent
+      }
+
+      private def mkSubPreCalls(overrideNumRef: js.VarRef,
+          isConstructor: Set[String])(implicit pos: Position) = {
+        val overrideNumss = subConstructors.map(_.overrideNumBounds)
+        val paramRefs = getParamRefs
+        val bodies = subConstructors.map { cg =>
+          cg.mkPreSuperCallOnSndCtr(overrideNumRef, paramRefs, isConstructor)
+        }
+        overrideNumss.zip(bodies).foldRight[js.Tree](js.Skip()) {
+          case ((numBounds, body), acc) =>
+            val cond = mkOverrideNumsCond(overrideNumRef, numBounds)
+            js.If(cond, body, acc)(jstpe.BooleanType)
+        }
+      }
+
+      private def mkPreSuperCallOnSndCtr(
+          overrideNumRef: js.VarRef, outputParams: List[js.VarRef],
+          isConstructor: Set[String])(implicit pos: Position): js.Tree = {
+        val subCalls =
+          mkSubPreCalls(overrideNumRef, isConstructor)
+
+        val preSuperCall = {
+          method.body match {
+            case js.Block(stats) =>
+              val beforeSuperCall = stats.takeWhile {
+                case js.ApplyStatic(_, mtd, _)
+                    if ir.Definitions.isConstructorName(mtd.name) =>
+                  false
+
+                case _ => true
+              }
+              val superCallParams = stats.collectFirst {
+                case js.ApplyStatic(_, mtd, js.This() :: args)
+                    if ir.Definitions.isConstructorName(mtd.name) =>
+                  zipMap(outputParams, args) { (ref, tree) =>
+                      js.Assign(ref, tree)
+                  }
+              }.getOrElse(Nil)
+
+              beforeSuperCall ::: superCallParams
+
+            case js.ApplyStatic(_, mtd, js.This() :: args)
+                if ir.Definitions.isConstructorName(mtd.name) =>
+              zipMap(outputParams, args)(js.Assign(_, _))
+
+            case _ => Nil
+          }
+        }
+
+        js.Block(subCalls :: preSuperCall)
+      }
+
+      private def mkSubPostCalls(overrideNumRef: js.VarRef,
+          isConstructor: Set[String])(implicit pos: Position): js.Tree = {
+        val overrideNumss = subConstructors.map(_.overrideNumBounds)
+        val bodies = subConstructors.map { cg =>
+          cg.mkPostSuperCallOnSndCtr(overrideNumRef, isConstructor)
+        }
+        overrideNumss.zip(bodies).foldRight[js.Tree](js.Skip()) {
+          case ((numBounds, js.Skip()), acc) => acc
+
+          case ((numBounds, body), acc) =>
+            val cond = mkOverrideNumsCond(overrideNumRef, numBounds)
+            js.If(cond, body, acc)(jstpe.BooleanType)
+        }
+      }
+
+      private def mkPostSuperCallOnSndCtr(overrideNumRef: js.VarRef,
+          isConstructor: Set[String])(implicit pos: Position): js.Tree = {
+        val postSuperCall = {
+          method.body match {
+            case js.Block(stats) =>
+              stats.dropWhile {
+                case js.ApplyStatic(_, mtd, _) if isConstructor(mtd.name) => false
```

Comment:
This seems to be the only place where `isConstructor: Set[String]` is actually used. Consider replacing it with `isConstructorName`:
```scala
case js.ApplyStatic(_, mtd, js.This() :: _) if isConstructorName(mtd.name) => false
```